src/goto-harness/recursive_initialization.cpp

```diff
@@ -774,7 +779,28 @@ code_blockt recursive_initializationt::build_struct_constructor(
   for(const auto &component :
       ns.follow_tag(to_struct_tag_type(struct_type)).components())
   {
-    initialize(member_exprt{dereference_exprt{result}, component}, depth, body);
+    if(component.get_is_padding())
+    {
+      continue;
+    }
+    // if the struct component is const we need to cast away the const
+    // for initialisation purposes.
+    // As far as I'm aware that's the closest thing to a 'correct' way
+    // to initialize dynamically allocated structs with const components
+    exprt component_initialisation_lhs = [&result, &component]() -> exprt {
+      auto member_expr = member_exprt{dereference_exprt{result}, component};
+      if(component.type().get_bool(ID_C_constant))
+      {
+        return dereference_exprt{
+          typecast_exprt{address_of_exprt{std::move(member_expr)},
```

Comment:
Comment that this pile of stuff amounts to `*(X*)&field`, where `field` has type `const X*`

Reply:
Also I take it you've tried using braced initialiser style for such things?